dynamic-ssl/src/main/java/org/wildfly/security/dynamic/ssl/DynamicSslSocketFactory.java

```diff
@@ -0,0 +1,148 @@
+
+    private DynamicSSLContextSPI dynamicSSLContextImpl;
+    private volatile String[] intersectionCipherSuite;
+    private SSLSocketFactory configuredDefaultSslSocketFactory;
+
+    DynamicSslSocketFactory(SSLSocketFactory configuredDefaultSslSocketFactory, DynamicSSLContextSPI dynamicSSLContextImpl) {
+        super();
+        Objects.requireNonNull(configuredDefaultSslSocketFactory);
+        Objects.requireNonNull(dynamicSSLContextImpl);
+        this.configuredDefaultSslSocketFactory = configuredDefaultSslSocketFactory;
+        this.dynamicSSLContextImpl = dynamicSSLContextImpl;
+    }
+
+    @Override
+    public Socket createSocket() throws IOException {
+        return configuredDefaultSslSocketFactory.createSocket();
+    }
+
+    @Override
+    public Socket createSocket(InetAddress address, int port) throws IOException {
+        return createSocketBasedOnPeerInfo(null, port, address, null, null, null, null);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port) throws IOException {
+        return createSocketBasedOnPeerInfo(host, port, null, null, null, null, null);
+    }
+
+    @Override
+    public Socket createSocket(String host, int port, InetAddress localAddress, int localPort) throws IOException {
+        return createSocketBasedOnPeerInfo(host, port, null, localAddress, localPort, null, null);
+    }
+
+    @Override
+    public Socket createSocket(InetAddress address, int port, InetAddress localAddress, int localPort) throws IOException {
+        return createSocketBasedOnPeerInfo(null, port, address, localAddress, localPort, null, null);
+    }
+
+    @Override
+    public Socket createSocket(Socket socket, String host, int port, boolean autoClose) throws IOException {
+        return createSocketBasedOnPeerInfo(host, port, null, null, null, socket, autoClose);
+    }
+
+    @Override
+    public String[] getDefaultCipherSuites() {
+        return configuredDefaultSslSocketFactory.getDefaultCipherSuites();
+    }
+
+    @Override
+    public String[] getSupportedCipherSuites() {
+        String[] val = intersectionCipherSuite;
+        if (val == null) {
+            synchronized (this) {
+                val = intersectionCipherSuite;
+                if (intersectionCipherSuite == null) {
+                    val = intersectionCipherSuite = getIntersection();
+                }
+            }
+        }
+        return val;
+    }
+
+    private Socket createSocketBasedOnPeerInfo(String hostname, Integer port, InetAddress address, InetAddress localAddress, Integer localPort, Socket socket, Boolean autoClose) throws IOException {
+        try {
+            SSLContext sslContext = this.dynamicSSLContextImpl.getSSLContext(new URI(null, null, hostname == null ? address.getHostName() : hostname, port, null, null, null));
+            if (sslContext == null) {
+                throw new IllegalStateException("Received SSLContext from DynamicSSLContextProvider was null");
```

Comment:
not sure if I should throw exception here or return configured default ssl context or jvm default ssl context